README.md
MobileApp
=========
2016-12-29
    升级到spring-boot1.4
    
2016-12-27
    增加druid监控；
    http://127.0.0.1:9006/form/rest/druid/index.html
    
2016-12-20
    菜单调整
    http://192.168.1.119:9006/form/rest/resource/search/findAllByPidOrderByModuleAsc?pid=resource
    
2016-11-29
    完成java 动态脚本调用groovy json 示例,作为flume 插件的规则脚本使用。
    
    
2016-10-28
    增加extjs 文件浏览示例
    http://127.0.0.1:9006/form/rest/public/file-browser.html
    http://127.0.0.1:9006/form/rest/logviewer/files
    
2016-10-18
    菜单编辑界面
    http://127.0.0.1:9006/form/rest/public/treegrid.html
    
    
2016-10-14
    zTree 树形目录；
    http://127.0.0.1:9006/form/rest/html/ResourceTree.html
    http://127.0.0.1:9006/form/rest/resource/search/findAllByOrderByModuleAsc
    http://127.0.0.1:9006/form/rest/resource/search/findTop10ByOrderByModuleDesc
    

2016-09-21
    增加mvc 拦截器；
    
2016-09-09
    spring-boot增加session保存Redis
    spring-boot-redis 集群测试 OK.
    
    
2016-08-19
    支持domain 多包扫描；
    增加部分注解；
    
    
2016-08-05
    注销mongo 的异步调用；
    注销redis and mongo 的调用；
    //@Configuration
    @PropertySource("classpath:application.properties")
    public class MongoConfiguration {
    
    //@Configuration
    public class RedisConfiguration {
    
    用户管理：
    http://127.0.0.1:9006/form/rest/user
    
2016-07-21
    domain 元数据
    http://127.0.0.1:9006/form/rest/webmetas?domainName=com.supermy.security.domain.Resource
    

2016-07-20
    增加资源数据
    权限数据管理；
    用户管理
    角色管理
    群组管理
    权限管理
    资源管理
    
    
2016-07-19
    增加健康控制：
    http://127.0.0.1:9006/form/rest/health
    http://127.0.0.1:9006/form/rest/trace
            <dependency>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-starter-actuator</artifactId>
            </dependency>
            
    更新测试方案
        @RunWith(SpringJUnit4ClassRunner.class)
        @SpringApplicationConfiguration(classes = Application.class)
        @WebAppConfiguration
        public class MongoDBGeoSpatialTest {
        
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
     
    ------------------------------------
    com.supermy.geo已经移植到push-server;代码暂不删除。
    
    完成rbac 权限模型样例数据
    
    
    
2016-07-18
    todo:
        rbac 数据准备与api 准备；
        
    微服务基础框架准备已经完成。
    
    extjs-filter 已经移植到es-query;详见web+app;
    清除原有filter的代码改造 ok；
    升级spring-boot1.2.7 ok;
    升级spring-boot1.3.5 ok;
    
    
    
2016-07-14
    
    {"DATA":
        [{"deleted":null,"enabled":null,"createDate":"2016-07-14T00:00:00.000+0800","createBy":"","updateDate":"2016-07-14T09:55:50.000+0800","updateBy":"","name":"xyz","code":"xyz","pwd":"\u79d8\u94a5","tokenExpire":90,"iplist":"bb","ipBindtime":300,"ipTimeout":60,"connectCount":100,"limitBandwidth":"3000","status":false,
            "pkId":4,"_links":{"self":{"href":"http://172.16.71.56:9006/form/rest/channel_auth/filter/channel_auth/4","templated":false}},"_embedded":{}},
         {"deleted":null,"enabled":null,"createDate":"2016-07-14T00:00:00.000+0800","createBy":"system1","updateDate":"2016-07-14T09:48:29.000+0800","updateBy":"system2","name":"test","code":"test","pwd":"test","tokenExpire":1,"iplist":"192.168.59.103","ipBindtime":300,"ipTimeout":61,"connectCount":100,"limitBandwidth":"100","status":true,
            "pkId":1,"_links":{"self":{"href":"http://172.16.71.56:9006/form/rest/channel_auth/filter/channel_auth/1","templated":false}},"_embedded":{}},
         {"deleted":null,"enabled":null,"createDate":"2016-07-14T00:00:00.000+0800","createBy":"","updateDate":"2016-07-14T09:55:46.000+0800","updateBy":"","name":"abc","code":"abc","pwd":"\u79d8\u94a5","tokenExpire":90,"iplist":"aa","ipBindtime":300,"ipTimeout":60,"connectCount":100,"limitBandwidth":"2000","status":false,
            "pkId":3,"_links":{"self":{"href":"http://172.16.71.56:9006/form/rest/channel_auth/filter/channel_auth/3","templated":false}},"_embedded":{}}],
     "PAGE":{"size":23,"totalElements":3,"totalPages":1,"number":0}}
     
     {"DATA":
        {"_links":{},"_embedded":{"channel_auth":[{"deleted":null,"code":"test","pkId":1,"pwd":"test","connectCount":100,"enabled":null,"limitBandwidth":"100","ipBindtime":300,"updateBy":"system2","updateDate":"2016-07-14T09:48:29.000+0800","createDate":"2016-07-14T00:00:00.000+0800","tokenExpire":1,"_links":{"self":{"href":"http:\/\/172.16.71.56\/form\/rest\/channel_auth\/1","templated":false}},"name":"test","_embedded":{},"iplist":"192.168.59.103","createBy":"system1","status":true,"ipTimeout":61}]}},
     "PAGE":{"page":{"totalElements":30}}}
     
     
    extjs 要求的返回值
    更新：单个记录与总记录数量
    删除：result response:......{"DATA":{},"PAGE":{"page":{"totalElements":2}}}

2016-07-13
    参数数组查询使用in金勋查询；
    http://127.0.0.1:9006/form/rest/channel_auth/search/findByIdList?pkId=1&pkId=2
    http://172.16.71.56:9006/form/rest/channel_auth/search/findByIdList?pkId=1&pkId=2



2016-07-08
    ？？？fixme 只是filter 通过es-query,返回id-list,再通过rest-idlist 查询返回数据
    ***idList 多个同名参数传递获得数组，可以使用hql 的in 语法进行查询；
    http://127.0.0.1:9006/form/rest/channel_auth/search/findByCodeList?code=test&code=abc
    
    
2016-07-07
    rest-api 负责增删改；
    es 负责查询以及全文检索；
    spark-sql 负责大数据查询；
    
    http://127.0.0.1:9006/form/rest/channel_auth/filter?_dc=1467884593669&sort=name%2CASC&sort=code%2CDESC&page=0&filter=%5B%7B%22type%22%3A%22string%22%2C%22value%22%3A%22test%22%2C%22field%22%3A%22name%22%7D%2C%7B%22type%22%3A%22string%22%2C%22value%22%3A%22test%22%2C%22field%22%3A%22code%22%7D%5D&start=0&size=23
    
    _dc=1467884593669
    sort=name,ASC
    sort=code,DESC
    page=0
    filter=[{"type":"string","value":"test","field":"name"},
            {"type":"string","value":"test","field":"code"}]
    start=0
    size=23
    
    
2016-07-06
    转移到nginx-lua 项目
    extjs filter 参数
    _dc:1467776408383
    sort:name,ASC
    sort:code,DESC
    page:0
    filter:[{"type":"string","value":"12","field":"name"},{"type":"string","value":"abc1","field":"code"}]
    start:0
    size:23
    
    走extjs[get]->nginx-lua[post]->直接发送到 es;
    {
    "query": {
      "bool": {
        "must": {//必须符合以下条件
          "term": {
            "name": "test"
          }
        },
        "should": {//满足其中某个条件
          "term": {
            "name": "test"
          }
        },
        "must_not": {//不能符合以下条件
          "term": {
            "code": "test1"
          }
        }
      }
    },
    "from": 0, //分页
    "size": 1,
    "sort": {  //排序
      "code": {
        "order": "desc"
      }
    }
    }
    
    
2016-06-20
    todo:
    基于extjs 的查询从spring-boot 的filter 移植到es;
    
2016-06-18
    增加scala 编译环境
    mvn spring-boot:run
    time curl -v http://127.0.0.1:9006/form/rest/helloscala
    
2016-05-28
    此模块后续主要处理非实时的业务功能交互;
    查询附近的司机\乘客网络约车等功能已经移植到GeoEngine 项目;
    
2016-05-25
    mvn dependency:tree

    指令采用json 格式传递,不采用POJO 序列化;
    redis+lua 进行抢红包/秒杀/抢单的核心实现;
    
    
    
    字符串解析成JavaBean：当数据量较少时首选FastJson，数据量较大使用Jackson。但是Jackson无法堆一个对象集合进行解析，只能转成一个Map集合，这点Gson和FastJson处理的比较好。
    字符串解析成JSON：当数据量较少时首选FastJson，数据量较大使用Jackson。
    JavaBean构造JSON：当数据量较少时选择Gson，数据量较大可使用Jackson。
    集合构造JSON：首先Jackson，其次Fastjson。

    todo 优化 com.fasterxml.jackson.core 的应用
    
2016-05-24
    引入spring-redis 框架

    匹配订单,向司机发送订单信息完成.
    

    订单信息格式
    o:流水号
    
        > db.location.save( {_id: "18610586581", position: [0.1, -0.1]} )
        > db.location.save( {_id: "18610586582", position: [1.0, 1.0]} )
        > db.location.save( {_id: "18610586583", position: [0.5, 0.5]} )
        > db.location.save( {_id: "18610586584", position: [-0.5, -0.5]} )
         
        接着指定location索引
        > db.location.ensureIndex( {position: "2d"} )
        
2016-05-23
    先实现逻辑,快速开发;
    然后迭代改造,加上nginx+mq;
    
    通过单ip 访问频率,控制用户访问频率;
    
    
2016-05-18
    http://127.0.0.1:9006/form/rest/webgeo/circle?x1=0.1&y1=0.1&bj=1
    mongodb 加上验证;
    
    todo:条件组合查询使用es 进行改造;
    

2016-05-16
    ConnectionString uri = new ConnectionString("mongodb://user1:pwd1@host1/?authSource=db1");


    geo LBS 服务圆圈查询
    http://127.0.0.1:9006/form/rest/webgeo/circle?x1=0.0&y1=0.0&bj=0.7
    http://127.0.0.1:9006/form/rest/webgeo/circle?x1=0.0&y1=0.0&bj=0.75
    http://127.0.0.1:9006/form/rest/webgeo/near?x1=0.1&y1=0.1&bj=700
    http://127.0.0.1:9006/form/rest/webgeo/near?x1=0.1&y1=0.1&bj=750
    
    准备数据
    首先定义一个位置集合,给定a,b,c,d节点.
    > db.createCollection("location")
    { "ok" : 1 }
    > db.location.save( {_id: "A", position: [0.1, -0.1]} )
    > db.location.save( {_id: "B", position: [1.0, 1.0]} )
    > db.location.save( {_id: "C", position: [0.5, 0.5]} )
    > db.location.save( {_id: "D", position: [-0.5, -0.5]} )
     
    接着指定location索引
    db.location.ensureIndex( {position: "2d"} )
    现在我们可以进行简单的GEO查询
    
    查询point(0,0),半径0.7附近的点
    > db.location.find( {position: { $near: [0,0], $maxDistance: 0.7  } } )
    { "_id" : "A", "position" : [ 0.1, -0.1 ] }
     
    查询point(0,0),半径0.75附近的点 
    > db.location.find( {position: { $near: [0,0], $maxDistance: 0.75  } } )
    { "_id" : "A", "position" : [ 0.1, -0.1 ] }
    { "_id" : "C", "position" : [ 0.5, 0.5 ] }
    { "_id" : "D", "position" : [ -0.5, -0.5 ] }
     
    我们可以看到半径不一样,查询出的点也不一样,因为c点坐标为[0.5,0.5],c至圆点的距离根据勾股定理可得出Math.sqrt(0.25 +0.25) ≈ 0.707,所以最大距离0.7时查找不到你要的点.
    
    查询[0.25, 0.25], [1.0,1.0]区域附近的点
    > db.location.find( {position: { $within: { $box: [ [0.25, 0.25], [1.0,1.0] ] }  } } )
    { "_id" : "C", "position" : [ 0.5, 0.5 ] }
    { "_id" : "B", "position" : [ 1, 1 ] }
     
    
2016-05-12
    插入mongodb 一条记录;
    curl -i -X POST -H "Content-Type:application/json" -d '{  "name" : "张三",  "sex" : "男" }' http://127.0.0.1:9006/form/rest/customer

CHANGELOG.md
MobileApp
=========
2017-04-20
    maven rocksdb 要与 brew list rocksdb 版本一致
    
2017-01-16
    去掉neo4j 的引用；
    单独项目处理；
    回退到jdk1.7
    
2016-12-29
    升级到spring-boot1.4
    neo4j 使用自带的rest 进行图数据库的维护与引擎的调度；
    
    
2016-12-27
    增加druid监控；
    http://127.0.0.1:9006/form/rest/druid/index.html
    
2016-12-20
    菜单调整
    http://192.168.1.119:9006/form/rest/resource/search/findAllByPidOrderByModuleAsc?pid=resource
    
2016-11-29
    完成java 动态脚本调用groovy json 示例,作为flume 插件的规则脚本使用。
    
    
2016-10-28
    增加extjs 文件浏览示例
    http://127.0.0.1:9006/form/rest/public/file-browser.html
    http://127.0.0.1:9006/form/rest/logviewer/files
    
2016-10-18
    菜单编辑界面
    http://127.0.0.1:9006/form/rest/public/treegrid.html
    
    
2016-10-14
    zTree 树形目录；
    http://127.0.0.1:9006/form/rest/html/ResourceTree.html
    http://127.0.0.1:9006/form/rest/resource/search/findAllByOrderByModuleAsc
    http://127.0.0.1:9006/form/rest/resource/search/findTop10ByOrderByModuleDesc
    

2016-09-21
    增加mvc 拦截器；
    
2016-09-09
    spring-boot增加session保存Redis
    spring-boot-redis 集群测试 OK.
    
    
2016-08-19
    支持domain 多包扫描；
    增加部分注解；
    
    
2016-08-05
    注销mongo 的异步调用；
    注销redis and mongo 的调用；
    //@Configuration
    @PropertySource("classpath:application.properties")
    public class MongoConfiguration {
    
    //@Configuration
    public class RedisConfiguration {
    
    用户管理：
    http://127.0.0.1:9006/form/rest/user
    
2016-07-21
    domain 元数据
    http://127.0.0.1:9006/form/rest/webmetas?domainName=com.supermy.security.domain.Resource
    

2016-07-20
    增加资源数据
    权限数据管理；
    用户管理
    角色管理
    群组管理
    权限管理
    资源管理
    
    
2016-07-19
    增加健康控制：
    http://127.0.0.1:9006/form/rest/health
    http://127.0.0.1:9006/form/rest/trace
            <dependency>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-starter-actuator</artifactId>
            </dependency>
            
    更新测试方案
        @RunWith(SpringJUnit4ClassRunner.class)
        @SpringApplicationConfiguration(classes = Application.class)
        @WebAppConfiguration
        public class MongoDBGeoSpatialTest {
        
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
     
    ------------------------------------
    com.supermy.geo已经移植到push-server;代码暂不删除。
    
    完成rbac 权限模型样例数据
    
    
    
2016-07-18
    todo:
        rbac 数据准备与api 准备；
        
    微服务基础框架准备已经完成。
    
    extjs-filter 已经移植到es-query;详见web+app;
    清除原有filter的代码改造 ok；
    升级spring-boot1.2.7 ok;
    升级spring-boot1.3.5 ok;
    
    
    
2016-07-14
    
    {"DATA":
        [{"deleted":null,"enabled":null,"createDate":"2016-07-14T00:00:00.000+0800","createBy":"","updateDate":"2016-07-14T09:55:50.000+0800","updateBy":"","name":"xyz","code":"xyz","pwd":"\u79d8\u94a5","tokenExpire":90,"iplist":"bb","ipBindtime":300,"ipTimeout":60,"connectCount":100,"limitBandwidth":"3000","status":false,
            "pkId":4,"_links":{"self":{"href":"http://172.16.71.56:9006/form/rest/channel_auth/filter/channel_auth/4","templated":false}},"_embedded":{}},
         {"deleted":null,"enabled":null,"createDate":"2016-07-14T00:00:00.000+0800","createBy":"system1","updateDate":"2016-07-14T09:48:29.000+0800","updateBy":"system2","name":"test","code":"test","pwd":"test","tokenExpire":1,"iplist":"192.168.59.103","ipBindtime":300,"ipTimeout":61,"connectCount":100,"limitBandwidth":"100","status":true,
            "pkId":1,"_links":{"self":{"href":"http://172.16.71.56:9006/form/rest/channel_auth/filter/channel_auth/1","templated":false}},"_embedded":{}},
         {"deleted":null,"enabled":null,"createDate":"2016-07-14T00:00:00.000+0800","createBy":"","updateDate":"2016-07-14T09:55:46.000+0800","updateBy":"","name":"abc","code":"abc","pwd":"\u79d8\u94a5","tokenExpire":90,"iplist":"aa","ipBindtime":300,"ipTimeout":60,"connectCount":100,"limitBandwidth":"2000","status":false,
            "pkId":3,"_links":{"self":{"href":"http://172.16.71.56:9006/form/rest/channel_auth/filter/channel_auth/3","templated":false}},"_embedded":{}}],
     "PAGE":{"size":23,"totalElements":3,"totalPages":1,"number":0}}
     
     {"DATA":
        {"_links":{},"_embedded":{"channel_auth":[{"deleted":null,"code":"test","pkId":1,"pwd":"test","connectCount":100,"enabled":null,"limitBandwidth":"100","ipBindtime":300,"updateBy":"system2","updateDate":"2016-07-14T09:48:29.000+0800","createDate":"2016-07-14T00:00:00.000+0800","tokenExpire":1,"_links":{"self":{"href":"http:\/\/172.16.71.56\/form\/rest\/channel_auth\/1","templated":false}},"name":"test","_embedded":{},"iplist":"192.168.59.103","createBy":"system1","status":true,"ipTimeout":61}]}},
     "PAGE":{"page":{"totalElements":30}}}
     
     
    extjs 要求的返回值
    更新：单个记录与总记录数量

src/main/webapp/public/common-lua.js
//设置插件路径
//var webserver="http://127.0.0.1:9006/form/rest";
//nginx+lua
var webserver="http://192.168.99.101/extjsform/proxy";
var formserver='http://192.168.99.101/extjsform';

var abc=tttt;


var pageSize=23;

Ext.Loader.setConfig({enabled: true});
Ext.Loader.setPath('Ext.ux', '../ux');

Ext.Loader.setPath('Ext.ux', 'ext4.2/ux');

Ext.require([
    'Ext.grid.*',
    'Ext.data.*',
    'Ext.ux.grid.FiltersFeature',
    'Ext.toolbar.Paging',
    'Ext.ux.ajax.JsonSimlet',
    'Ext.ux.ajax.SimManager'
]);

//Ext.require([
//    '*',
//    'Ext.toolbar.Paging',
//    'Ext.ux.grid.FiltersFeature',//必不可少的
//    'Scripts.*'
//]);
//
//var filters = {
//    alias: 'widget.filters',
//    ftype: 'filters',
//
//    encode: false
//    //, // json encode the filter query
//
//    //指定要对哪些列进行过滤
//    //filters: [{
//    //    type: 'boolean',
//    //    dataIndex: 'IsSuccessed'
//    //}]
//};


//修正chrome 浏览器下extjs grid filter 的过滤的编辑面板不好用的问题
// fix hide submenu (in chrome 43)
Ext.override(Ext.menu.Menu, {
    onMouseLeave: function (e) {
        var me = this;


        // BEGIN FIX
        var visibleSubmenu = false;
        me.items.each(function (item) {
            if (item.menu && item.menu.isVisible()) {
                visibleSubmenu = true;
            }
        })
        if (visibleSubmenu) {
            //console.log('apply fix hide submenu');
            return;
        }
        // END FIX


        me.deactivateActiveItem();


        if (me.disabled) {
            return;
        }


        me.fireEvent('mouseleave', me, e);
    }
});

//验证错误不能变回问题，重写RowEditor解决问题
Ext.override(Ext.grid.RowEditor,
    {
        addFieldsForColumn: function (column, initial) {
            var me = this, i, length, field;
            if (Ext.isArray(column)) {
                for (i = 0, length = column.length; i < length; i++) {
                    me.addFieldsForColumn(column[i], initial);
                }
                return;
            }
            if (column.getEditor) {
                field = column.getEditor(null, {
                    xtype: 'displayfield',
                    getModelData: function () {
                        return null;
                    }
                });
                if (column.align === 'right') {
                    field.fieldStyle = 'text-align:right';
                }
                if (column.xtype === 'actioncolumn') {
                    field.fieldCls += ' ' + Ext.baseCSSPrefix + 'form-action-col-field';
                }
                if (me.isVisible() && me.context) {
                    if (field.is('displayfield')) {
                        me.renderColumnData(field, me.context.record, column);
                    } else {
                        field.suspendEvents();
                        field.setValue(me.context.record.get(column.dataIndex));
                        field.resumeEvents();
                    }
                }
                if (column.hidden) {
                    me.onColumnHide(column);
                } else if (column.rendered && !initial) {
                    me.onColumnShow(column);
                }

                // -- start edit
                me.mon(field, 'change', me.onFieldChange, me);
                // -- end edit
            }
        }
    });



Date.createFromMysql = function(mysql_string)
{
    var t, result = null;

    if( typeof mysql_string === 'string' )
    {
        t = mysql_string.split(/[- :]/);

        //when t[3], t[4] and t[5] are missing they defaults to zero
        result = new Date(t[0], t[1] - 1, t[2], t[3] || 0, t[4] || 0, t[5] || 0);
    }

    return result;
}

//对后端返回的json 进行标准化处理和封装；
function processResponse(success, operation, request, response, callback, scope) {

    //自定义--begin ,列表的时候初始化，新增和删除的时候读取total 数据
    console.log("process response:......" ,this.model.total);

    console.log("process responseText:......" ,response.responseText);
    console.log("process request:......" ,request.url);
    console.log("process operation:......" ,operation);
    console.log("process success:......" , success);

    //var mystrjson = response.responseText;
    var myobjson = Ext.decode("{}");
    if (!Ext.isEmpty(response.responseText)){
        var myobjson = Ext.decode(response.responseText);//将json字符串转换为对象
    }

    if (Ext.isEmpty(myobjson["DATA"])) {
        console.log("lua 没有进行filter 处理 ...........");

        //后台请求成功,处理返回数据
        if (success) {

            //var mystrjson2;
            var myobjson2 = Ext.decode("{}");
            myobjson2["DATA"] = myobjson;

            var page;
            //对后端返回的json 进行标准化处理和封装；
            if (request.action == 'read') {

                myobjson2["DATA"] = {};
                if (!Ext.isEmpty(myobjson._embedded)) {
                    //方法1:使用正则根据filter 特征获取数据的标签;
                    //var s = request.url;
                    //var dl = new RegExp("[a-zA-z0-9]+/(filter)", "i").exec(s)[0];
                    //var datalabel = /[a-zA-z0-9]+/.exec(dl)[0];
                    //
                    //console.log("datalabel:......", datalabel);
                    //console.log(myobjson._embedded[datalabel]);
                    //myobjson2["DATA"] = myobjson._embedded[datalabel];

                    //方法2:获取json key;利用json的数据层次获取数据
                    for (var key in myobjson._embedded) {
                        console.log(key);
                        console.log(myobjson._embedded[key]);
                        myobjson2["DATA"] = myobjson._embedded[key];
                    }
                }

                page = myobjson.page;
            }
            else if (request.action == 'create') {
                this.model.total = this.model.total + 1;

                page = {
                    "page": {
                        "totalElements": this.model.total
                    }
                };

            }

            else if (request.action == 'update') {
                page = {
                    "page": {
                        "totalElements": this.model.total
                    }
                };
            }

            else if (request.action == 'destroy') {
                this.model.total = this.model.total - 1;
                page = {
                    "page": {
                        "totalElements": this.model.total
                    }
                };
            }

            myobjson2["PAGE"] = page;

            //mystrjson2 = Ext.encode(myobjson2);//将json对象转换为json字符串 其中Mystrjson=Mystrjson2
            console.info("result response:......" + Ext.encode(myobjson2));//使用json对象属性

            response.responseText = Ext.encode(myobjson2);//将json对象转换为json字符串 其中Mystrjson=Mystrjson2

        } else {//后台处理失败,提示错误信息;debug = true 显示详细错误信息,显示概要描述信息.

            Ext.MessageBox.alert("操作失败", myobjson.message, function (id) {
                //var userStore = operation.getStore();
                var userStore = Ext.getStore('Users');
                userStore.reload();
            });

        }
    }

    //自定义--end

    var me = this,
        reader,
        result;

    if (success === true) {
        reader = me.getReader();

        // Apply defaults to incoming data only for read operations.
        // For create and update, there will already be a client-side record
        // to match with which will contain any defaulted in values.
        reader.applyDefaults = operation.action === 'read';

        result = reader.read(me.extractResponseData(response));


        if (result.success !== false) {
            //see comment in buildRequest for why we include the response object here
            Ext.apply(operation, {
                response: response,
                resultSet: result
            });

            operation.commitRecords(result.records);
            operation.setCompleted();
            operation.setSuccessful();
        } else {
            operation.setException(result.message);
            me.fireEvent('exception', this, response, operation);
        }
    } else {
        me.setException(operation, response);
        me.fireEvent('exception', this, response, operation);
    }

    //this callback is the one that was passed to the 'read' or 'write' function above
    if (typeof callback == 'function') {
        callback.call(scope || me, operation);
    }

    me.afterRequest(request, success);
}




function doRequest (operation, callback, scope) {
    console.log("do request:", operation);

    var writer = this.getWriter(),
        request = this.buildRequest(operation);
    if (operation.allowWrite()) {
        request = writer.write(request);
    }


    Ext.apply(request, {
        async: this.async,
        binary: this.binary,
        headers: this.headers,
        timeout: this.timeout,
        scope: this,
        callback: this.createRequestCallback(request, operation, callback, scope),
        method: this.getMethod(request),
        disableCaching: false
    });


    /*
     * do anything needed with the request object
     */
    //console.log('request', request);
    console.log('request.params', request.params);
    //对filter 参数进行加工处理  fixme 或者后台处理
    //var filterstr = new Array();
    //function filterArrayElements(element, index, array) {
    //    console.log("filter element:"+element.field);
    //    filterstr[index] = element.field + "=" + element.value;
    //}
    //request.params.filter.forEach(filterArrayElements);
    //
    //var params = {
    //    filter: filterstr
    //};
    //Ext.apply(operation.getstore().proxy.extraParams, params);


    Ext.Ajax.request(request);
    return request;
}

//4store 提交后处理
function afterRequest (request, success) {
    //var store=Ext.getStore("channelsid");
    //统一给用户修改提示信息
    console.log("after request:", request);

    //var mystrjson = response.responseText;
    //var mystrjson2;
    //var myobjson = Ext.decode(mystrjson);//将json字符串转换为对象


    var result = request.operation.success;

    if (request.action == 'read') {

        //记录记录总数，标准化返回参数。
        if (!Ext.isEmpty(request.operation.resultSet)){
            this.model.total = request.operation.resultSet.total;
            console.log("记录总数:",  this.model.total);
        }



        Ext.MessageBox.show({
            msg: '正在查询数据,请稍后...',
            progressText: 'Saving...',
            width: 300,
            wait: true,
            waitConfig: {interval: 200},
            icon: 'ext-mb-download', //custom class in msg-box.html
            iconHeight: 50
        });
        setTimeout(function () {
            //This simulates a long-running operation like a database save or XHR call.
            //In real code, this would be in a callback function.
            Ext.MessageBox.hide();
        }, 500);
    }

    else if (request.action == 'create') {
        if (result) {
            Ext.MessageBox.show({
                msg: '正在保存数据,请稍后...',
                progressText: 'Saving...',
                width: 300,
                wait: true,
                waitConfig: {interval: 200},
                icon: 'ext-mb-download', //custom class in msg-box.html
                iconHeight: 50
            });
            setTimeout(function () {
                //This simulates a long-running operation like a database save or XHR call.
                //In real code, this would be in a callback function.
                Ext.MessageBox.hide();
            }, 500);

            //Ext.Msg.alert('添加提示', '添加成功！');
            //store.reload();
        }
        //else {
        //    Ext.Msg.alert('添加提示', '添加失败！');
        //}
    }

    else if (request.action == 'update') {
        if (result) {
            Ext.MessageBox.show({
                msg: '正在保存数据,请稍后...',
                progressText: 'Saving...',
                width: 300,
                wait: true,
                waitConfig: {interval: 200},
                icon: 'ext-mb-download', //custom class in msg-box.html
                iconHeight: 50
            });
            setTimeout(function () {
                //This simulates a long-running operation like a database save or XHR call.
                //In real code, this would be in a callback function.
                Ext.MessageBox.hide();
            }, 500);
            //Ext.Msg.alert('提示', '更新成功！');
            //store.reload();
        }
        //else {
        //    Ext.Msg.alert('提示', '更新失败！');
        //}
    }

    else if (request.action == 'destroy') {
        if (result) {
            Ext.MessageBox.show({
                msg: '正在删除数据,请稍后...',
                progressText: 'Delete...',
                width: 300,
                wait: true,
                waitConfig: {interval: 200},
                icon: 'ext-mb-download', //custom class in msg-box.html
                iconHeight: 50
            });
            setTimeout(function () {
                //This simulates a long-running operation like a database save or XHR call.
                //In real code, this would be in a callback function.
                Ext.MessageBox.hide();
            }, 500);
            //Ext.Msg.alert('提示', '数据删除成功');
            //store.reload();
        }
        //else {
        //    //Ext.Msg.alert('提示', '数据删除失败');
        //}
    }
}



function metachange  (store, meta) {
    console.log("Version " + meta.version + "Search query " + meta.searchQuery);
}

//查询参数预处理
function beforeload (store, operation, eOpts) {
    console.log(store);
    console.log(operation);

    //排序参数预处理
    var sortstr = new Array();
    function sortArrayElements(element, index, array) {
        sortstr[index] = element.property + "," + element.direction;
    }
    operation.sorters.forEach(sortArrayElements);


    //将搜索条件放在store的baseParams中，每次加载都赋值
    //var new_params = { name: Ext.getCmp('search').getValue() };
    //Ext.apply(store.proxy.extraParams, "searchkeyword");
    //var params = {sthecstr:conditionstr};
    //Ext.apply(options.params, params);
    //store.baseParams = {limit: gridPageSize, filter: $('#txtFilter').val()};


    //翻页参数预处理，以及排序参数更新
    var params = {
        sort: sortstr,
        page: store.currentPage - 1
    };
    Ext.apply(store.proxy.extraParams, params);



}


//JS时区转载函数
//function formatTime(date){
//    //CST时间和UTC时差转换
//    var formatPattern = 'Y-m-d H:i:s';
//
//    var dateStr=date.toString();
//    var position=dateStr.indexOf("UTC");
//    if(position!=-1){
//        return Ext.util.Format.date(date, formatPattern);
//    }else{
//        console.log('here......');
//        var vdate=new Date(date); //后台默认为格林尼治时间，该方法会默认把后台时间认为是北京时间，北京时间比格林尼治时间多14小时。
//        //找出当地时间偏移值的毫秒数
//        var localOffset=vdate.getTimezoneOffset()*60000;
//        var utcOffset= vdate.getTime() + localOffset;
//        var timezone=utcOffset+8*3600000;
//        //var timezone=utcOffset;
//
//        var lastDate=new Date(timezone);
//        return Ext.util.Format.date(lastDate, formatPattern);
//    }
//}

//JS时区转载函数
//function formatDate(date){
//    //CST时间和UTC时差转换
//    var formatPattern = 'Y-m-d';
//
//    var dateStr=date.toString();
//    var position=dateStr.indexOf("UTC");
//    if(position!=-1){
//        return Ext.util.Format.date(date, formatPattern);
//    }else{
//        console.log('here......');
//        var vdate=new Date(date); //后台默认为格林尼治时间，该方法会默认把后台时间认为是北京时间，北京时间比格林尼治时间多14小时。
//        //找出当地时间偏移值的毫秒数
//        var localOffset=vdate.getTimezoneOffset()*60000;
//        var utcOffset= vdate.getTime() + localOffset;
//        var timezone=utcOffset+8*3600000;
//        //var timezone=utcOffset;
//
//        var lastDate=new Date(timezone);
//        return Ext.util.Format.date(lastDate, formatPattern);
//    }
//}

//function convertDiffTime(date){
//        console.log('here......'+date);
//        var formatPattern = 'Y-m-d H:i:s';
//
//        var vdate=new Date(date); //后台默认为格林尼治时间，该方法会默认把后台时间认为是北京时间，北京时间比格林尼治时间多14小时。
//        //找出当地时间偏移值的毫秒数
//        var localOffset=vdate.getTimezoneOffset()*60000;
//        var utcOffset= vdate.getTime() + localOffset;
//        var timezone=utcOffset+8*3600000;
//        //var timezone=utcOffset;
//
//        var lastDate=new Date(timezone);
//    console.log(lastDate);
//
//    console.log(Ext.util.Format.date(lastDate, formatPattern));
//
//        return timezone;
//
//}
//
//function convertDiffDate(date){
//    console.log('here......');
//    var formatPattern = 'Y-m-d H:i:s';
//
//    var vdate=new Date(date); //后台默认为格林尼治时间，该方法会默认把后台时间认为是北京时间，北京时间比格林尼治时间多14小时。
//    //找出当地时间偏移值的毫秒数
//    var localOffset=vdate.getTimezoneOffset()*60000;
//    var utcOffset= vdate.getTime() + localOffset;
//    var timezone=utcOffset+8*3600000;
//    //var timezone=utcOffset;
//
//    var lastDate=new Date(timezone);
//    console.log(Ext.util.Format.date(lastDate, formatPattern));
//
//    return lastDate;
//
//}
